Added Swagger generation

elasti-ron opened the pull request:
Hi,

I've added the capability to auto-generate static [Swagger](https://swagger.io/) definition files from the Apipie DSL.

The resulting Swagger ([OpenAPI 2.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md)) compliant JSON file can then be used to auto-generate REST clients, imported into tools like Postman, and so forth.

Swagger generation is implemented by crawling through Apipie's resource_descriptions construct.

The DSL is slightly enhanced to allow indication of default values of optional parameters.

There are a few configuration options that can be used to change the overall structure of the generated swagger file, in order to support both in:formData and in:body REST request styles.

Looking forward to hear your thoughts.

iNecas commented on lib/apipie/swagger_generator.rb:
```diff
@@ -0,0 +1,583 @@
+module Apipie
+  module Validator
+    class EnumValidator < BaseValidator
```
Why not putting it directly to the validator definition?

iNecas commented on lib/apipie/swagger_generator.rb:
```diff
@@ -0,0 +1,583 @@
+    end
+
+
+    def save_and_compare
```
This method is not being called from any place? Should it be exposed from the rake?

iNecas commented on lib/apipie/swagger_generator.rb:
```diff
@@ -0,0 +1,583 @@
+    # note that at the moment, this only takes operation ids into account, and ignores parameter
+    # definitions, so it's only partially useful.
+    def include_op_id_in_master_identifier(op_id)
+      @master_id = Zlib::crc32("#{@master_id} #{op_id}")
```
we should probably require `zlib` somewhere: it's probably loaded by rails somewhere, but just to be on safe side

iNecas commented on lib/apipie/swagger_generator.rb:
```diff
@@ -0,0 +1,583 @@
+      # print "??? RESEARCH: [#{ruby_name_for_method(@current_method)}] -- #{msg}\n"
+    end
+
+    def dump_param_desc(param_desc)
```
Unused method?

iNecas commented:
There are some unused methods and commented code, that should be cleaned first. I have not tested this yet, but my view on this is we could get this in pretty soon, as it has not influence on other features, and can be pretty useful for many folks.

iNecas commented:
I wonder if we could have a test on checking that the swagger output valid, in the tests.

iNecas commented:
Another question: would it make sense to expose this via an `/apidoc` endpoint, similarly as we have with `apidoc.json`?

elasti-ron commented:
Thanks @iNecas  - good points.  I'll push a cleaned-up version early next week.

Re exposing via `/apidoc`:  yes, would be useful.  Would probably make sense to auto-fill the swagger `host` field with the hostname/port from the request object.  I'll look into this.

Re testing the validity of the swagger output: would be great to add. A possible approach might be to validate against the [Swagger schema](https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json) (perhaps by using the [json-shema](https://github.com/ruby-json-schema/json-schema) gem?) on the output. What do you think?

iNecas commented:
Agreed with all you said :)

elasti-ron commented:
Pushed a cleaned-up version addressing the code issues you highlighted.

Will keep you posted on progress re the `/apidoc` endpoint and the swagger schema validation.

elasti-ron commented:
Hi @iNecas - it turns out that rendering at runtime from `/apidoc` is not as simple as I had hoped.  It will take me some time before I'm able to look into this further, and I'll only be able to continue working on this (and on the swagger schema validation) next week.

If you feel comfortable with what I've posted so far as a first version, it might be best to merge that into the master branch and not wait for the additional functionality.  This will reduce the risk of incompatibilities creeping into the code.

iNecas commented:
I'm ok with waiting for the `/apidoc` endpoint, but I wukd lje to have the schema validator test before merging this to keep it working with the further updates

elasti-ron commented:
The schema validation was surprisingly easy to add.  Pushed a version that includes it.

elasti-ron commented:
ping

elasti-ron commented:
Hi @iNecas - runtime swagger generation is now supported (with some limitations - see the README).

iNecas commented on app/controllers/apipie/apipies_controller.rb:
```diff
@@ -14,11 +14,43 @@ def authenticate
+
+      @language = get_language
+
+      Apipie.load_documentation if Apipie.configuration.reload_controllers? || (Rails.version.to_i >= 4.0 && !Rails.application.config.eager_load)
```
There is quite a lot of code duplicated from the `index` method. Could you look into how to DRY this a bit?

Otherwise, the implementation looks good and I think we are close to merging this

iNecas commented on lib/apipie/configuration.rb:
```diff
@@ -165,6 +170,12 @@ def initialize
+      @swagger_content_type_input = :form_data  # this can be :json or :form_data
+      @swagger_json_input_uses_refs = false
+      @swagger_include_warning_tags = false
+      @swagger_suppress_warnings = false #[105,100,102]
```
What does ` #[105,100,102]` mean?

iNecas replied:
Nm: I looked at the readme and here we go :)

iNecas commented:
Thanks @elasti-ron, I don't see any other reason why holding this in a PR any longer. Thanks for hard work and this one. I'm sure it will be useful for others as well

elasti-ron commented:
Thanks @iNecas - it was good to collaborate with you on this.

As a next step, I'd like to add support for describing response structures, as it would make the output much more useful.

I've written a [proposal](https://gist.github.com/elasti-ron/e6cb880fdbcf855ef22143021a435bbf) to describe my suggested approach, and I would highly appreciate your feedback on it.

iNecas commented:
Thanks for moving apipie forward. I would have few comments to the proposal: would you mind opening the proposal as a PR, with the markdown document in it. The would for the PR would not be to merge that, but to have a meaningful discussion over the diff itself.

In general, I like the idea of combining the concepts we already have in the apipie. Although, I would suggest introducing additional keyword (something like `property`) to the dsl: when defining response only keys, I'm not sure it makes sense to talk about `parameter` in that regard. It would be more like an alias.

I will have more comments, but I would like to leave that for a new PR to continue with the discussion.

iNecas commented:
Also, make sure to advertise this in https://github.com/Apipie/apipie-rails/issues/273, as there are other folks interested into this functionality, that could have a valuable input to the progress of this.

elasti-ron commented:
Done.  #588.
